File: DataStructureJava/LinkedListQueue.java
/**
 * 使用链表实现的队列
 * 这里我们为链表添加了两个指针，分别指向链表头和链表尾部
 * 在链表头部，我们添加和删除元素仅需要O(1)的时间复杂度
 * 在链表尾部，由于新加了指向尾部的tail指针，我们可以方便的添加元素，但是如果我们要删除尾节点，那我们必须通过遍历获取到尾节点的前一个节点，时间复杂度为O(n)
 * 综合考虑以上情况我们可以将链表的头部充当队列的头部，只做删除元素
 * 链表的尾部充当队列的尾部，只添加元素，这样就可以实现入队和出队都是O(1)的时间复杂度 
 */
public class LinkedListQueue<E> implements Queue<E> {
    private class Node {
        public E e;
        public Node next;

        public Node(E e, Node next) {
            this.e = e;
            this.next = next;
        }

        public Node(E e) {
            this(e, null);
        }

        public Node() {
            this(null, null);
        }

        @Override
        public String toString() {
            return e.toString();
        }
    }

    private Node head, tail;
    private int size;

    public LinkedListQueue() {
        head = null;
        tail = null;
        size = 0;
    }

    @Override
    public int getSize() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public void enqueue(E e) {
        if (tail == null) {
            tail = new Node(e);
            head = tail;
        } else {
            tail.next = new Node(e);
            tail = tail.next;
        }
        size++;
    }

    @Override
    public E dequeue() {
        if (isEmpty())
            throw new IllegalArgumentException("Cannot dequeue from an empty queue.");
        Node retNode = head;
        head = retNode.next;
        retNode.next = null;
        if (head == null)
            tail = null;
        size--;
        return retNode.e;
    }

    @Override
    public E getFront() {
        if (isEmpty())
            throw new IllegalArgumentException("Queue is empty");
        return head.e;
    }

    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();
        res.append("Queue: front");
        Node cur = head;
        while (cur != null) {
            res.append(cur + "->");
            cur = cur.next;
        }
        res.append("NULL tail");
        return res.toString();
    }

    public static void main(String[] args) {
        LinkedListQueue<Integer> queue = new LinkedListQueue<>();
        for(int i = 0 ; i < 10 ; i ++){
            queue.enqueue(i);
            System.out.println(queue);

            if(i % 3 == 2){
                queue.dequeue();
                System.out.println(queue);
            }
        }
    }

}

File: DataStructureTypeScript/LinkedListQueue.ts
import Queue from './interface/Queue';
import LinkedListNode from './common/LinkedListNode';

export default class LinkedListQueue<E> implements Queue<E> {
  public head: LinkedListNode<E> = null;
  private tail: LinkedListNode<E> = null;
  private size: number = 0;

  /**
   * 获取队列中当前存储元素的个数
   * Time Complexity O(1)
   * @return {number}
   */
  getSize() {
    return this.size;
  }

  /**
   * 判断队列是否为空，即元素的个数是否为0
   * Time Complexity O(1)
   * @return {boolean}
   */
  isEmpty(): boolean {
    return this.size === 0;
  }

  /**
   * 向队列末尾添加一个元素
   * Time Complexity O(1)
   * @param {E} e 要添加的元素
   * @return {void}
   */
  enqueue(e: E): void {
    if (this.tail === null) {
      this.tail = new LinkedListNode<E>(e);
      this.head = this.tail;
    } else {
      this.tail.next = new LinkedListNode<E>(e);
      this.tail = this.tail.next;
    }
    this.size++;
  }

  /**
   * 从队列头部出队一个元素，并返回出队的元素
   * Time Complexity O(n)
   * @return {E}
   */
  dequeue(): E {
    if (this.isEmpty()) {
      throw new Error('Cannot dequeue from an empty queue.');
    }
    let retNode = this.head;
    this.head = this.head.next;
    retNode.next = null; // 切断之前的头结点和头结点下一个节点之间的联系
    if (this.head === null) {
      this.tail = null;
    }
    this.size--;
    return retNode.e;
  }

  /**
   * 查看队列头部的元素
   * Time Complexity O(1)
   * @return {E}
   */
  getFront(): E {
    return this.head.e;
  }

  // 辅助方法：用于打印队列
  toString() {
    let res = '';
    res += 'Queue: front';
    let cur = this.head;
    while (cur !== null) {
      res += `${cur}->`;
      cur = cur.next;
    }
    res += 'NULL tail';
    return res;
  }
}
